File: back-end-java/src/main/java/com/aie/backend/modules/project/service/ProjectService.java
package com.aie.backend.modules.project.service;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.aie.backend.modules.auth.entities.Administrador;
import com.aie.backend.modules.auth.security.AuthenticatedAdmin;
import com.aie.backend.modules.project.dto.CreateProjectRequest;
import com.aie.backend.modules.project.dto.ProjectShareRequest;
import com.aie.backend.modules.project.dto.UpdateProjectRequest;
import com.aie.backend.modules.project.entities.Project;
import com.aie.backend.modules.project.entities.ProjectShare;
import com.aie.backend.modules.project.repository.ProjectRepository;
import com.aie.backend.modules.project.repository.ProjectShareRepository;
import com.aie.backend.modules.response.repository.ResponseRepository;

@Service
@Transactional
public class ProjectService {

    private static final Logger LOGGER = LoggerFactory.getLogger(ProjectService.class);

    private final ProjectRepository repository;
    private final ProjectShareRepository shareRepository;
    private final ResponseRepository responseRepository;

    public ProjectService(ProjectRepository repository,
                          ProjectShareRepository shareRepository,
                          ResponseRepository responseRepository) {
        this.repository = repository;
        this.shareRepository = shareRepository;
        this.responseRepository = responseRepository;
    }

    public Project create(CreateProjectRequest request, AuthenticatedAdmin user) {
        Long ownerId = user != null ? user.id() : null;
        if (ownerId == null) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.forbidden");
        }
        Project project = new Project();
        project.setName(request.getName().trim());
        project.setResponsible(request.getResponsible().trim());
        project.setDescription(Optional.ofNullable(request.getDescription()).orElse(null));
        project.setLastPretriagemLevel(request.getLastPretriagemLevel());
        project.setLastPretriagemScore(request.getLastPretriagemScore() != null ? BigDecimal.valueOf(request.getLastPretriagemScore()) : null);
        project.setActive(true);
        project.setOwner(new Administrador() {{
            setId(ownerId);
        }});

        Project saved = repository.save(project);
        saved.setIsOwner(true);
        saved.setSharedWithMe(false);
        saved.setResponsesCount(0L);
        return saved;
    }

    @Transactional(readOnly = true)
    public List<Project> findAll(AuthenticatedAdmin user) {
        if (user == null || user.id() == null) {
            return List.of();
        }
        List<Project> projects = repository.findAccessible(user.id(), user.socialNumber());
        Map<Long, Long> finishedCounts = getFinishedCounts(projects);

        for (Project project : projects) {
            boolean isOwner = project.getOwner() != null && user.id().equals(project.getOwner().getId());
            boolean sharedWithMe = !isOwner && project.getOwner() != null && project.getOwner().getId() != null;
            project.setIsOwner(isOwner);
            project.setSharedWithMe(sharedWithMe);
            project.setResponsesCount(finishedCounts.getOrDefault(project.getId(), 0L));
        }
        return projects;
    }

    @Transactional(readOnly = true)
    public Project findOne(Long id, AuthenticatedAdmin user) {
        Project project = repository.findByIdAndActiveTrue(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "project.not_found"));
        ensureAccess(project, user, true);
        boolean isOwner = project.getOwner() != null && user.id().equals(project.getOwner().getId());
        boolean sharedWithMe = !isOwner && project.getOwner() != null && project.getOwner().getId() != null;
        project.setIsOwner(isOwner);
        project.setSharedWithMe(sharedWithMe);
        project.setResponsesCount(getFinishedCount(project.getId()));
        return project;
    }

    public Project update(Long id, UpdateProjectRequest request, AuthenticatedAdmin user) {
        Project project = ensureOwnerAccess(id, user);
        if (request.getName() != null) project.setName(request.getName().trim());
        if (request.getResponsible() != null) project.setResponsible(request.getResponsible().trim());
        if (request.getDescription() != null) project.setDescription(request.getDescription());
        if (request.getLastPretriagemLevel() != null) project.setLastPretriagemLevel(request.getLastPretriagemLevel());
        if (request.getLastPretriagemScore() != null) project.setLastPretriagemScore(BigDecimal.valueOf(request.getLastPretriagemScore()));

        Project saved = repository.save(project);
        saved.setIsOwner(true);
        saved.setSharedWithMe(false);
        saved.setResponsesCount(getFinishedCount(saved.getId()));
        return saved;
    }

    public Map<String, Boolean> remove(Long id, AuthenticatedAdmin user) {
        Project project = ensureOwnerAccess(id, user);
        project.setActive(false);
        repository.save(project);
        return Map.of("success", Boolean.TRUE, "softDeleted", Boolean.TRUE);
    }

    @Transactional(readOnly = true)
    public List<ProjectShare> listShares(Long projectId, AuthenticatedAdmin user) {
        ensureOwnerAccess(projectId, user);
        return shareRepository.findByProjectIdOrderByDateCreatedAsc(projectId);
    }

    public ProjectShare addShare(Long projectId, ProjectShareRequest request, AuthenticatedAdmin user) {
        Project project = ensureOwnerAccess(projectId, user);
        String social = request.getSocialNumber().trim();
        if (social.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.share_invalid");
        }

        if (project.getOwner() != null && project.getOwner().getSocialNumber() != null) {
            if (project.getOwner().getSocialNumber().equals(social)) {
                throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.share_invalid_owner");
            }
        }

        return shareRepository.findByProjectIdAndSocialNumber(projectId, social)
                .orElseGet(() -> {
                    ProjectShare share = new ProjectShare();
                    share.setProject(project);
                    share.setSocialNumber(social);
                    return shareRepository.save(share);
                });
    }

    public Map<String, Boolean> removeShare(Long projectId, Long shareId, AuthenticatedAdmin user) {
        ensureOwnerAccess(projectId, user);
        ProjectShare share = shareRepository.findById(shareId)
                .filter(s -> s.getProject() != null && projectId.equals(s.getProject().getId()))
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "project.share_not_found"));
        shareRepository.delete(share);
        return Map.of("success", Boolean.TRUE);
    }

    private Project ensureOwnerAccess(Long projectId, AuthenticatedAdmin user) {
        if (user == null || user.id() == null) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.forbidden");
        }
        Project project = repository.findByIdAndActiveTrue(projectId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "project.not_found"));
        if (project.getOwner() == null || !user.id().equals(project.getOwner().getId())) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.forbidden");
        }
        return project;
    }

    private void ensureAccess(Project project, AuthenticatedAdmin user, boolean allowShared) {
        if (user == null || user.id() == null) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.forbidden");
        }
        if (project.getActive() != null && !project.getActive()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "project.not_found");
        }
        boolean isOwner = project.getOwner() != null && user.id().equals(project.getOwner().getId());
        boolean shared = false;
        if (allowShared && project.getShares() != null && user.socialNumber() != null) {
            shared = project.getShares().stream().anyMatch(s -> user.socialNumber().equals(s.getSocialNumber()));
        }
        if (!isOwner && !shared) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "project.forbidden");
        }
    }

    private Map<Long, Long> getFinishedCounts(List<Project> projects) {
        Map<Long, Long> map = new HashMap<>();
        List<Long> ids = projects.stream().map(Project::getId).toList();
        if (ids.isEmpty()) return map;
        responseRepository.countByProjectIdsAndStatus(ids, "FINISHED").forEach(row -> map.put(row.getProjectId(), row.getTotal()));
        return map;
    }

    private long getFinishedCount(Long projectId) {
        List<Long> ids = List.of(projectId);
        return responseRepository.countByProjectIdsAndStatus(ids, "FINISHED").stream()
                .findFirst().map(row -> row.getTotal()).orElse(0L);
    }
}

File: back-end-nestjs/src/modules/project/project.service.ts
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
// src/modules/project/project.service.ts
import {
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Brackets, Repository } from 'typeorm';
import { Project } from './entities/project.entity';
import { CreateProjectDto } from './dto/create-project.dto';
import { UpdateProjectDto } from './dto/update-project.dto';
import { ProjectShare } from './entities/project-share.entity';
import { ProjectShareDto } from './dto/project-share.dto';

@Injectable()
export class ProjectService {
  constructor(
    @InjectRepository(Project)
    private readonly repo: Repository<Project>,
    @InjectRepository(ProjectShare)
    private readonly shareRepo: Repository<ProjectShare>,
  ) {}

  async create(dto: CreateProjectDto, user: any) {
    const project = this.repo.create({
      ...dto,
      owner: user?.id ? ({ id: user.id } as any) : null,
    });
    return this.repo.save(project);
  }

  async findAll(user: any) {
    if (!user?.id) return [];
    const qb = this.repo
      .createQueryBuilder('project')
      .leftJoinAndSelect('project.owner', 'owner')
      .leftJoin('project.shares', 'share')
      .loadRelationCountAndMap(
        'project.responses_count',
        'project.responses',
        'finishedResponses',
        (qb) => qb.where('finishedResponses.status = :status', { status: 'FINISHED' }),
      )
      .where('project.owner_id = :ownerId', { ownerId: user.id });

    if (user.social_number) {
      qb.orWhere('share.social_number = :social', {
        social: user.social_number,
      });
    }

    qb.andWhere('project.active = :active', { active: true });

    const projects = await qb
      .distinct(true)
      .orderBy('project.id', 'DESC')
      .getMany();

    return projects.map((project) => ({
      ...project,
      is_owner: project.owner_id === user.id,
      shared_with_me:
        project.owner_id != null && project.owner_id !== user.id ? true : false,
    }));
  }

  async findOne(id: number, user: any) {
    const project = await this.findProjectWithAccess(id, user);
    if (!project) throw new NotFoundException('project.not_found');
    return project;
  }

  async update(id: number, dto: UpdateProjectDto, user: any) {
    await this.ensureOwnerAccess(id, user);
    await this.repo.update(id, dto);
    return this.repo.findOne({ where: { id } });
  }

  async remove(id: number, user: any) {
    await this.ensureOwnerAccess(id, user);
    await this.repo.update(id, { active: false });
    return { success: true, softDeleted: true };
  }

  async listShares(projectId: number, user: any) {
    await this.ensureOwnerAccess(projectId, user);
    return this.shareRepo.find({
      where: { project: { id: projectId } },
      order: { date_created: 'ASC' },
    });
  }

  async addShare(projectId: number, dto: ProjectShareDto, user: any) {
    const project = await this.ensureOwnerAccess(projectId, user);
    const social = dto.social_number?.trim();
    if (!social) throw new ForbiddenException('project.share_invalid');

    if (
      project.owner &&
      project.owner.social_number &&
      project.owner.social_number === social
    ) {
      throw new ForbiddenException('project.share_invalid_owner');
    }

    const exists = await this.shareRepo.findOne({
      where: {
        project: { id: project.id },
        social_number: social,
      },
    });
    if (exists) return exists;

    const share = this.shareRepo.create({
      project,
      social_number: social,
    });
    return this.shareRepo.save(share);
  }

  async removeShare(projectId: number, shareId: number, user: any) {
    await this.ensureOwnerAccess(projectId, user);
    const share = await this.shareRepo.findOne({
      where: { id: shareId, project: { id: projectId } },
    });
    if (!share) throw new NotFoundException('project.share_not_found');
    await this.shareRepo.delete(share.id);
    return { success: true };
  }

  private async findProjectWithAccess(id: number, user: any) {
    if (!user?.id) throw new ForbiddenException('project.forbidden');
    const qb = this.repo
      .createQueryBuilder('project')
      .leftJoinAndSelect('project.owner', 'owner')
      .leftJoin('project.shares', 'share')
      .loadRelationCountAndMap(
        'project.responses_count',
        'project.responses',
        'finishedResponses',
        (qb) => qb.where('finishedResponses.status = :status', { status: 'FINISHED' }),
      )
      .where('project.id = :id', { id })
      .andWhere(
        new Brackets((qbWhere) => {
          qbWhere.where('project.owner_id = :ownerId', { ownerId: user.id });
          if (user.social_number) {
            qbWhere.orWhere('share.social_number = :social', {
              social: user.social_number,
            });
          }
        }),
      )
      .andWhere('project.active = :active', { active: true })
      .distinct(true);

    return qb.getOne();
  }

  private async ensureOwnerAccess(projectId: number, user: any) {
    if (!user?.id) throw new ForbiddenException('project.forbidden');
    const project = await this.repo.findOne({
      where: { id: projectId, active: true },
      relations: ['owner'],
    });
    if (!project) throw new NotFoundException('project.not_found');
    if (project.owner_id !== user.id) {
      throw new ForbiddenException('project.forbidden');
    }
    return project;
  }
}
